---
title: "01_DESEQ_counts"
author: "JR"
date: "8/9/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)

# Final re-installation of everything...

# packageurl <- "https://github.com/tidyverse/tidyverse/archive/refs/tags/v1.3.1.tar.gz"
# install.packages(packageurl, repos=NULL, type="source")
# 
# if (!requireNamespace("BiocManager", quietly = TRUE))
# install.packages("BiocManager", version = '3.10')
# 
# BiocManager::install("GenomicRanges",dependencies = TRUE, INSTALL_opts = '--no-lock')
# 
# install.packages("https://cran.r-project.org/src/contrib/Archive/locfit/locfit_1.5-9.4.tar.gz", repos=NULL, type="source")
# install.packages("https://cran.r-project.org/src/contrib/Archive/latticeExtra/latticeExtra_0.6-29.tar.gz", repos=NULL, type="source")
# install.packages("https://cran.r-project.org/src/contrib/Archive/Hmisc/Hmisc_3.6-0.tar.gz", repos=NULL, type="source")
# 
# BiocManager::install("DESeq2")


# now load the package
library(tidyverse)
library(GenomicRanges)
library(DESeq2)

```
Accompanying YouTube Video:
https://youtu.be/czZoz4JUWdE


Today we will be exploring the results of our NF_CORE rnaseq pipeline run.
We noted a mistake in encode labeling :) so we will fix that and create a final 
samplesheet for differential expression analyses by DESEQ in the next class.

GOAL: to understand how to convert the output counts from Salmon to rlog normalized counts.
We will also generate all the needed files we need to do differential expression analyses with DESEQ.

It is always good to look at the raw data (bigWig in IGV etc)

# We discovered ENCODE swapped the labels on two samples !!
This was discovered two differnet ways that are good starting points:

(1) NF_core multiQC showed that two different experiment types were 
more similar than their replicates. Most samples cluster by experimental_accession.
For example: 

ENCSR813BDU_rep1 (insoluble_cytoplasmic_fraction) & ENCSR181ZGR_rep1 (homo_sapiens_hepg2) 
so did:
ENCSR813BDU_rep2 (insoluble_cytoplasmic_fraction) & ENCSR181ZGR_rep2 (homo_sapiens_hepg2)
This is suspicious as we would expect the same experimental type to cluster together.

(2) By looking in the browser at the raw data.
We focused on a well known nuclear RNA : NEAT1. This gene should
not be expressed in "insolube_cytoplasmic" and was. Thus, one of the
insoluble fractions was total (homo_spaiens_hepg2) and the other insoluble_cytoplasmic.


We have concluded that:

# hepg2_R2 -- is whole cell / total
# hepg2_insoluble_cytoplasmic_fraction_R2 -- is whole cell / total
# hepg2_R1 -- is insoluble_cytoplasmic
# hepg2_insoluble_cytoplasmic_fraction_R1 -- is insoluble_cytoplasmic


Let's start by fixing our sample_sheet!
# FIRST, we need copy our samplesheet and design file into our 01_differential_expression dir.
USe the terminal to "cp" the folder to the current working directory:
<your_folder>/CLASS_2023/CLASSES/05_R_analyses/analysis/05_RNAseq/01_differential_expression

# cp samplesheet.csv ../../01_differential_expression/results
# cp design.csv ../../01_differential_expression/results


```{r import sample sheet}

# First let's read in th sample sheet we made in 00_NF_core_pipeline
samplesheet <- read_csv("../00_RNAseq_download_NF_core_pipeline/00_NF_CORE_RNAseq_Pipeline_run/samplesheet.csv")

# First changing the total sample to homo_sapiens_insoluble_fraction
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_R1"), "condition"] <- "homo_sapiens_insoluble_cytoplasmic_fraction"

# same for changing insoluble_cytoplasmic_fraction condition to homo_sapiens_hepg2
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_insoluble_cytoplasmic_fraction_R2"), "condition"] <- "homo_sapiens_hepg2"


# rewriting sample name
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_R1"), "sample_name"] <- "homo_sapiens_hepg2_insoluble_cytoplasmic_fraction_RX"


# rewrite sample name
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_insoluble_cytoplasmic_fraction_R2"), "sample_name"] <- "homo_sapiens_hepg2_R1"

# change RX back
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_insoluble_cytoplasmic_fraction_RX"), "sample_name"] <- "homo_sapiens_hepg2_insoluble_cytoplasmic_fraction_R2"

# finally change replicate numbers -- NOT IDEAL!
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_R1"), "replicate"] <- "R1"

# finally change replicate numbers -- NOT IDEAL!
samplesheet[which(samplesheet$sample_name == "homo_sapiens_hepg2_insoluble_cytoplasmic_fraction_R2"), "replicate"] <- "R2"


# saving this to sample_sheets folder
write.csv(samplesheet, "results/corrected_samplesheet.csv")

samplesheet <- read_csv("results/corrected_samplesheet.csv")



```

# Loading in Gencode annotations & making gene_id to gene_symbol object 
Now let's read in the genome features also in our data folder
While we are at it, let's also print these out to make this go quicker in the future.
```{r import genome features and make gene_id and symbol file}

# Importing genome annotations (we will export what we need in .Rdata format to avoid this step moving forward)
gencode_gtf <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

# making genocde_genes GRanges to save in .Rdata session.
gencode_genes <- gencode_gtf[gencode_gtf$type == "gene"]

# It is really handy to convert gene_id (long number) to gene_name (more intuitive). 
# We are making G2S file that we can use in future to convert between the two.
g2s <- gencode_genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

# writingn this out to avoid loading Gencode in the future
write.csv(g2s, "results/g2s.csv")

```

Ok, next step is to read in GENE COUNTS (not TPM) -- raw counts over gene features.
This was generated in the NF_CORE RNAseq pipeline in the salmon folder. 


More on salmon transcript quants, this is a basic outline of how Salmon works 
Especially if you wanted to use it outside NF_CORE pipeline for RNAseq
https://combine-lab.github.io/salmon/getting_started/

Here is more on output files
https://salmon.readthedocs.io/en/latest/file_formats.html

# reading in salmon merged gene counts.
# We want "salmon_merged_gene_counts.csv

```{r reading in salmon counts convert to matrix}

# DEseq requires a matrix of raw counts so the following
# will turn this from a data.frame to matrix for input into DEseq.
## NOTE ## NEVER USE TPM or any other normalized counts with DeSeq!!

# reading in salmon counts:
counts <- read_csv("../00_RNAseq_download_NF_core_pipeline/00_NF_CORE_RNAseq_Pipeline_run/results/salmon/salmon_merged_gene_counts.csv")

# adding rownames and converting to a matrix
counts <- column_to_rownames(counts, "gene_id") %>%
as.matrix()
  
# Nice now we have a matrix we need to import to DeSeq
# However it has all the samples so we need to organize a bit.
```

# Connecting samplesheet to counts
Setting up counts object to be used as input into DEseq
to create "Rlog_Counts". Rlog_counts are very important in comparing 
samples. They are normalized by expression quantile to prevent low abundance
transcripts from inflating data. 

```{R sub-selecting counts_matrix to data only in samplesheet }

# First let's put the counts columns in the same order as the samplesheet
# If they are not then we would essentially be randomizing our data later!!
counts <- counts[,samplesheet$sample_id]

# This is a handy check point to make sure the labels in 
# sample sheet and counts are similar
all(colnames(counts) == samplesheet$sample_id)

# Now we know they're in the same order, and we can change the names
# It is nicer to have more descriptive names.
colnames(counts) <- samplesheet$sample_name

```

# creating and exporting normalized counts

DeSeq needs things like:
1) numbers to be rounded to an integer 
2) Add "factor" levels to conditions being compared
3) filter out any genes that don't have any counts

```{R Getting counts organized for DeSeq }

# Deseq wants round numbers so we are going to round them here:
counts <- round(counts)

# now we want to get rid of all genes with 0 across all samples.
counts_filtered <- counts[rowSums(counts) > 1,]

# In order to cache this and read it in later we'll
# save it as an RDS
write_rds(counts_filtered, "results/counts_matrix.rds")

```

# Running DeSeq to create rlog normalzied counts!
Although a lot of normalization has already occurred from
Salmon (where we got the counts) we now need to determine the 
distribution of expression values and normalize again to prevent
low abundance genes inflating significant differential expression. 


# goal: to get rlog normalized counts 
Let's officially run DeSeq2 these are the steps we will take

(1) create an "design" for Deseq as input to create a 
"dds" (DESeqDataSet) :: "DESeqDataSetFromMatrix".
This comes from sample sheet.

# --- sample sheet needs to be factored! ----

```{R factor sample sheet}
# A FACTOR LEVEL is critical for DEseq2 to know which samples is which
# We will take the condition from sample sheet to compare for differential expression
# In this case it doesn't matter which comes first the "control"
samplesheet$condition <- as.factor(samplesheet$condition)

# Now that the samplesheet is ready to used in DESeq2, 
# we'll write it out for future use.
write_rds(samplesheet, "results/final_samplesheet.rds")
```


(2) run deseq to create a dds (DESeqDataSet) 
this requires a matrix of rounded counts (made above)
::  DESeq(dds)

(3) normalize the counts to "rlog_counts" ::
rlog(dds, blind = TRUE)

# Important info on this here:
https://compbiocore.github.io/deseq-workshop-1/assets/deseq_workshop_1.html

(4) retreive rlog counts using :: assay(rlog_counts)

```{R essential set up to running DeSeq}

# (1) first run DESeq2 by creating a dds object.
# We use the function "DESeqDataSetFromMatrix"
# With parameters of countData, colData, design

dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              # this is our counts data we read in
                              colData = samplesheet,
                              # telling DeSeq what env variable to use for sample sheet
                              design = ~ condition)
                              # perhaps most important is "condition" is a factor in samplesheet 

# (2) run DESeq2 function on dds object
dds <- DESeq(dds)

# (3) Normalize counts (rlog)
# This basically is rank counts normalized to std error in replicates.
rlog_counts <- rlog(dds, blind = TRUE)

# (4) now we retrieve the values using the "assay" function that converts to rlog_counts)
rlog_counts_matrix <- assay(rlog_counts)

# Now we can write this out and START from here in the future.
# create new folder 'rlog_count_matrices'
write_rds(rlog_counts_matrix, "results/rlog_counts_all.rds")

```

Ooofda -- that was a lot of data wrangling .. let's save the key
aspects in .Rdata format to load in the next class !
# Saving key files for faster loading in the future

```{r saving key parts of environment as start point in future}

save(counts, counts_filtered, rlog, g2s, gencode_genes, rlog_counts, rlog_counts_matrix, samplesheet, file = "results/DESEQ_RLOG.RData")
     
    
# erase environment and reload!
load("results/DESEQ_RLOG.RData", verbose = T)

# that never get's old :) however one 
# NOTE is that this can propogate errors.
# if objects are loaded that are incorrect to make
# new objects that is a problem :) Just use caution!


```






