---
title: "03_clustering"
author: "JR"
date: "8/3/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(ggdendro)
library(GenomicRanges)

install.packages("pheatmap")
library(pheatmap)

```
Accompanying Youtube Video:
https://youtu.be/MH92yILbbuo


Today we are going to use hierarchical clustering to identify DBPs with similar binding patterns.

In brief, this involves calculating the distance (euclidian or spearman) between all pairwise 
comparisons of DBP binding profiles. We have made a matrix of 1 and 0 if a DBP is bound at a 
given genome annotation feature (e.g., promoter, gene-body). This is a binary clustering and 
if we cluster a continuous variable (e.g. RNAseq expression levels) then we may want to make 
other considerations for clustering.

We can use those vectors of 1 and 0 to determine how similar the binding profiles are.

Ok let's get started by grabbing the peak_occurrence_matrix from 14_2_our_data_in_ranges.

# loading in promoter peak occurence and convert to matrix
```{r loading in promoter_peak_occurence_df}

#let's look at the promoter peak occurence matrix (takes a min to load)
# read.table will make this a data.frame but will keep row names.
promoter_peak_occurence_matrix <- read.table("../01_peak_features/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# Converting to a matrix format for correlation analysis
promoter_peak_occurence_matrix <- as.matrix(promoter_peak_occurence_matrix)
# Ok we see this is a very large matrix of 1 and 0 if a DBP is bound to a promoter.
# We can use these 36,000+ vectors of genes with 483 values of bound (1) or unbound (0)

```

# calculating distnace matrix
Let's get started by finding the 'distance' between each promtoer vector.
We will make a distance matrix of 437 x437 and then use that for clustering.
```{r calculating distance matrix}

# First let's take a look at the matrix
# note the meta data
promoter_peak_occurence_matrix[1:2,19:24]
# let's see if there are row and column names:

# row.names
row.names(promoter_peak_occurence_matrix)
# column names
colnames(promoter_peak_occurence_matrix)

# now we can calculate the distance of DBP (we could also do this for promoters)
# let's do a small test first as this can 10 min to calculate 451x451!

# NOTE : Type of distance is important -- euclidean not good for binary measurements
test <- dist(promoter_peak_occurence_matrix[1:2,], method = "euclidean")
test

# There are a number of measures to calculate distance between two vectors and we can check those out in the documentation.
?dist


# let's see how this changes if we use binary as the distance method
# since this is a binary vector we will use a binary distance metric

test2 <- dist(promoter_peak_occurence_matrix[1:2,], method = "binary")
test2

# Now let's create distance matrix for each pairwise vector comparison for the whole matrix now.
peak_occurence_dist <- dist(promoter_peak_occurence_matrix, method = "binary")
peak_occurence_dist
```

# Clustering distance matrix

Now that we have that, we can do hierarchical clustering. In brief, this will cluster 
vectors that have small distances (high correlation), and will iterate that process until
everything is in the same cluster.

```{R hclust of distance matrix}
# let's take a look at hclust

?hclust

# Alright we are ready to cluster everything!
bin_hier <- hclust(peak_occurence_dist, method = "complete")
# that was incrediblty fast to get the correlation values !

# we now can plot the information in bin_hier
plot(bin_hier)

# let's take a look at what we got and see distance distribution
hist(bin_hier$height)

# To make this bigger, we can plot it as a pdf the syntax is:
# pdf() tell file path to save and properties then we plot
# finally we call dev.off to take the output into PDF instead of plots.

# We can order this before saving it
# Now let's order the DBPs 
bin_hier$labels[bin_hier$order]

# let's look at the order - environment shows smallest numbers first
# let's see the last one:
bin_hier$height[[4]]


# print out figure
#  !! CHECK WORKING DIRECTORY & structure !!
# Note we can't use ggsave here :( but this is a great way to save
# pdfs for all plots.

pdf("figures/dbp_hclust_dendro.pdf", height = 12, width = 70)
plot(bin_hier)
dev.off()

```


# ggdendro function for clustwering
Nice so we now have object "bin_hier" that will contain all the cross correlation values across all the samples. 

Now let's use GGDENDRO (dendrogram) package that will plot the branch lengths that indicate how similar two samples are 

```{r ggdendro gram}

# This is a lot of ggplot -- on purpose
# please read through what is happening and parameter usage

 ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
                       theme_dendro = TRUE) +
   # 90 degree rotation to right
   coord_flip() +
   scale_y_continuous() +
   # adds label
   scale_x_continuous(position = "top") +
   # subsection to labels in order of clustering
   # ? seq_along
   scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
                      
                      # adding labels that are in the order 'column'
             labels = bin_hier$labels[bin_hier$order], position = "top",
             expand = c(0,0)) +
   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
   scale_y_reverse(expand = c(0.01, 0)) +
   theme(
     plot.background = element_blank(),
     panel.grid.major = element_blank(),
   panel.grid.minor = element_blank(),
     panel.border = element_blank()
   )

# Nice let's save this using 'ggsave': since we used ggplot

ggsave("figures/ggdendro_plot.pdf", height = 50, width = 12, limitsize = F)

```


# Clustering of lncRNA and mRNA seperately
Now let's compare how lncRNAs and mRNAs cluster differnently?
Strategy: load annotations then index peak_occurrence for specific annotations

# lncRNA promoters ggdendro
```{r lncRNA promoter clustering}

# We first need to load the .GTF we made for lncRNAs and mRNAs.
# loading lncRNA promoters

lncrna_mrna_promoters <- rtracklayer::import("../01_peak_features/results/gene_annotations/lncrna_mrna_promoters.gtf")

# now we can split into lncRNA and mRNA
lncrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "lncRNA"]

# mrna promoters

mrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "protein_coding"]

# Now we will use indexing to separate peak_occurrence_matrix in lncRNA and mRNA.
# note we are indexing and using indexing in indexing to get what we want.
lncrna_peak_occurence <- promoter_peak_occurence_matrix[,lncrna_promoters$gene_id]

# we do the clutering the same as above or we can combine the dist and hclust:
bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))

# Now plot with ggdendro
ggdendro::ggdendrogram(bin_hier_lncrna, rotate = T,  size = 3)
 
# Now let's save this figure
ggsave("figures/lncrna_hclust_binary_dist.pdf", height = 49, width = 6)


```

# mRNA promter ggdendro
Now for mRNA

```{R mRNA promoter clustering}

# same strategy used very often:
# indexing into larger file and filtering to a file of stuff you want.
mrna_peak_occurence <- promoter_peak_occurence_matrix[,mrna_promoters$gene_id]

# getting the distance matrix for only mRNA promoters  
bin_hier_mrna <- hclust(dist(mrna_peak_occurence, method = "binary"))
 
# plotting with ggdendro
ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)

# saving
ggsave("figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)

# Cool we are making some good figures -- now let's dig into the biology
# Find an unknown protein cluster that has a known protein gene (hint ZNF)
# read about those genes and see if you could predict what the unknown gene is doing.

```

# Characterizing and Clustering "high binder promoters"
Let's make a heatmap of the promoters -- since it will take a long long time to plot we'll just plot the "super promoters"
```{r clustering high-binding promoters}
# first we filter promoter_peak_occurrence to just those that have high numbers of dbps
# let's, find the promoters that bind to many of our dbps
# a bit of trial and error but 5 of 7 dbps works for me.
high_binders <- promoter_peak_occurence_matrix[,colSums(promoter_peak_occurence_matrix) > 5]

# let's find out how many high binder promoters there are
ncol(high_binders)

# this is just a way to find those promoters where many proteins bind.
# we will use this much more with the full dataset of 480 DBPs
```

# Now let's make a heat map with pheatmap
Pheatmap will do the clustering and make a heatmap for us!
```{r heatmap}

# library(pheatmap)
?pheatmap

# Pheatmap wants a matrix of values, a method of clustering are the main parameters.
# This takes a minute !
pheatmap(promoter_peak_occurence_matrix, show_colnames = F, clustering_distance_rows = "binary", clustering_distance_cols = "binary")

# Save heat map.

pdf("figures/heatmap.pdf", height = 12, width = 12)
pheatmap(promoter_peak_occurence_matrix, show_colnames = F, clustering_distance_rows = "binary", clustering_distance_cols = "binary")
dev.off()

```

Excellent we can now make a heat map for all our DBPs and cluster them by 
similarity of what promoters they bind. We can already learn which proteins
are more similar in their binding to promoters!

#### Bonus material on what we can do with peak occurence matrix !!! ####
# Making a data frame of which proteins are bound on a given promoter 
Currently peak occurrence matrix just tells us whether a DBP overlapped 
a promoter. Promoters are cols and thus we can extract row name from
values in a col. 

To solve this we need to make a matrix that we can call a 
specific promoter and then find the DBPs bound to that promoter.

Thus you can simply look up your favorite promoter 
and retrieve all the DBPs bound to it.


# Step 1: make a dataframe of gene ids and gene_names
```{r DBPs bound to a given promoter}
# Let's make a data frame of gene_id and gene_name 
# we will merge this into our dbp_on_promoters below

# this is another way to make g2s file of gene_names and gene_symbols
DBPs_on_promoter <- bind_rows(lncrna_promoters %>% 
                                 as.data.frame(),
                 mrna_promoters %>% 
                    as.data.frame()) %>%
  dplyr::select(gene_id, gene_name)
```

# step 2: pivot longer and filter to only promoters a DBP binds
```{r DBPs bound to a given promoter}
# Here we will make a table to look up which DBPs are on a given promoter
promoter_dbps <- promoter_peak_occurence_matrix %>%
  as.data.frame() %>%
   # this adds "dbp" not as names but now a column
  rownames_to_column("dbp") %>%
# here we just made a column that has the DBP name
# currently it is a "meta" name
   # LET'S DISCUSS PIVOT WIDER and SHORTER - IT'S IMPORTANT
   # takes col 2: all the rest of cols (col 1 is dbp name)
   # now each DBP will have 30,000+ rows for values in all promoters
pivot_longer(2:ncol(.), names_to = "gene_id", values_to = "occurrence") %>%
# This now a very long matrix - let's get rid of all zero values
  filter(occurrence == 1) %>%
  dplyr::select(-occurrence) %>%
  left_join(DBPs_on_promoter)

# let's write this out it's a very handy DF!
# note saving in 12 where we read in this type of data
write.csv(promoter_dbps, "../01_peak_features/results/dbps_on_a_promoter.csv")
# This is a very large file but very useful you want to find which DBP is on which promoter

# write.csv(promoter_dbps, "/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/promoter_dbps.csv")

```

# Finding which DBPs are on which promoter
Cool now we can select any gene and see what DBPs are bound?

```{R DBPs on a given promoter}

# In the future if we want to filter on genes we can just read this back in.
promoter_dbps <- read.csv("../01_peak_features/results/dbps_on_a_promoter.csv")

# let's see what's on FIRRE
firre_promoter <- promoter_dbps %>%
  filter(gene_name == "FIRRE")

# seeing what's there :
firre_promoter
# How many DBPs bound
nrow(firre_promoter)

# GAPDH
GAPDH_promoter <- promoter_dbps %>%
  filter(gene_name == "GAPDH")

# seeing what's there :
GAPDH_promoter
# How many DBPs bound
nrow(GAPDH_promoter)

# trying to find a small binder :)
XIST_promoter <- promoter_dbps %>%
  filter(gene_name == "XIST")

# seeing what's there :
XIST_promoter
# How many DBPs bound
nrow(XIST_promoter)

# Interesting, intuition was correct considering
# this is a male cell line :)
```

################
ExeRcise
################

hint:

All you need to do is go into 01_peak_features and export lncrna_mrna_promoters with larger window

Repeat all code with promoter windows of size +3kb / -3kb

How do the clustering results change?

