---
title: "06_my_version"
author: "LVK"
date: "2/27/2023"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
# library(Gviz)
library(IRanges)
```

Accompanying Youtube video is here:
https://youtu.be/g_0Ylw-iHQ4

# genome features overlapping chip data
Now let's use GRanges to find overlaps of genome features and peaks from ChIPseq.
This is probably one of the most commonly used functions in genome sciences!
First we need to import POLR2A Chip-seq files -- a simple .bed file of peaks.

```{r reading in peaks}

# setting file path to peak files

basepath <- "/scratch/Shares/rinnclass/CLASS_2023/lucyvk"
path <- "CLASS_2023/CLASSES/03_Nextflow/00_my_chipseq/results/bwa/mergedLibrary/macs/broadPeak"

# first we read the peak files in as gRanges object with rtracklayer function.
peaks1 <- rtracklayer::import(file.path(basepath,path, "MAX_R1_peaks.broadPeak"))
peaks2 <- rtracklayer::import(file.path(basepath,path, "MAX_R2_peaks.broadPeak"))
peaks3 <- rtracklayer::import(file.path(basepath,path, "MAX_R3_peaks.broadPeak"))
peaks4 <- rtracklayer::import(file.path(basepath,path, "MAX_R4_peaks.broadPeak"))

# lengths of the peak files
length(peaks1) # 31522
length(peaks2) # 13048
length(peaks3) # 83146
length(peaks4) # 113543

# How well do my ChIPseq samples replicate? -- Look at overlap percentages between all the peaks files
# ov12 <-countOverlaps(peaks1,peaks2) # 12229
# sum(ov12)
# ov13 <-countOverlaps(peaks1,peaks3) # 29125
# sum(ov13)
# ov14 <-countOverlaps(peaks1,peaks4) # 29593
# sum(ov14)
# ov23 <-countOverlaps(peaks2,peaks3) # 12854
# sum(ov23)
# ov24 <-countOverlaps(peaks2,peaks4) # 12874
# sum(ov24)
# ov34 <-countOverlaps(peaks3,peaks4) # 74740
# sum(ov34)

# Create the tsv file versions and histograms!
MAX_R1_tsv <- read_tsv(file.path(basepath,path, "MAX_R1_peaks.broadPeak"), col_names = F)
MAX_R2_tsv <- read_tsv(file.path(basepath,path, "MAX_R2_peaks.broadPeak"), col_names = F)
MAX_R3_tsv <- read_tsv(file.path(basepath,path, "MAX_R3_peaks.broadPeak"), col_names = F)
MAX_R4_tsv <- read_tsv(file.path(basepath,path, "MAX_R4_peaks.broadPeak"), col_names = F)
names(MAX_R1_tsv) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 'signalValue', 'pValue', 'qValue')
names(MAX_R2_tsv) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 'signalValue', 'pValue', 'qValue')
names(MAX_R3_tsv) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 'signalValue', 'pValue', 'qValue')
names(MAX_R4_tsv) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 'signalValue', 'pValue', 'qValue')
MAX_R1_tsv <- filter(MAX_R1_tsv, chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"))
MAX_R2_tsv <- filter(MAX_R2_tsv, chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"))
MAX_R3_tsv <- filter(MAX_R3_tsv, chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"))
MAX_R4_tsv <- filter(MAX_R4_tsv, chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"))
MAX_R1_tsv <- MAX_R1_tsv %>% mutate(peak_width = end - start)
MAX_R2_tsv <- MAX_R2_tsv %>% mutate(peak_width = end - start)
MAX_R3_tsv <- MAX_R3_tsv %>% mutate(peak_width = end - start)
MAX_R4_tsv <- MAX_R4_tsv %>% mutate(peak_width = end - start)
h_MAX_R1 <- hist(MAX_R1_tsv$peak_width, breaks = 150, plot = FALSE)
h_MAX_R2 <- hist(MAX_R2_tsv$peak_width, breaks = 150, plot = FALSE)
h_MAX_R3 <- hist(MAX_R3_tsv$peak_width, breaks = 150, plot = FALSE)
h_MAX_R4 <- hist(MAX_R4_tsv$peak_width, breaks = 150, plot = FALSE)
plot(h_MAX_R1, xlim = c(0,2000), ylim = c(0,45000), xlab = "Peak Width", main = "MAX replicates peak width")
plot(h_MAX_R2, xlim = c(0,2000), border = "red", add = TRUE)
plot(h_MAX_R3, xlim = c(0,2000), border = "blue", add = TRUE)
plot(h_MAX_R4, xlim = c(0,2000), border = "green", add = TRUE)

# Load Gencode-v32: for genome features.
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

```

# promoters function
Now let's define the promoters using the handy promoters function in GRanges

```{R promoters}

# The promoter functions will use the first base in a genebody at the TSS.
# Then we can define the promoter up and downstream from there.
?promoters

# let's add 1Kb upstream and downstream from the TSS to define "promoters"
gencode_promoters <- promoters(gencode_gr[gencode_gr$type == "gene"], 
                               upstream = 1e3, 
                               downstream = 1e3)


length(gencode_promoters)
# there are ~60,000 promoters in the $type == gene

```

# How many peaks overlap promoters?
# findOVerlaps function
Now we have promoter ranges and peak ranges let's use findOverlaps function

```{r SubsetByOverlaps}

# replicate 1
promoter_overlaps_MAX_peaks1 <- subsetByOverlaps(peaks1, gencode_promoters) # 13801 out of 60609 - 22.8% of promoters
length(promoter_overlaps_MAX_peaks1)
length(promoter_overlaps_MAX_peaks1)/length(gencode_promoters)

# replicate 2:
promoter_overlaps_MAX_peaks2 <- subsetByOverlaps(peaks2, gencode_promoters) # 7786 out of 60609 - 12% of promoters
length(promoter_overlaps_MAX_peaks2)
length(promoter_overlaps_MAX_peaks2)/length(gencode_promoters)

# replicate 3:
promoter_overlaps_MAX_peaks3 <- subsetByOverlaps(peaks3, gencode_promoters) # 16571 - 27% of promoters
length(promoter_overlaps_MAX_peaks3)
length(promoter_overlaps_MAX_peaks3)/length(promoter_overlaps_MAX_peaks3)

# replicate 4:
promoter_overlaps_MAX_peaks4 <- subsetByOverlaps(peaks4, gencode_promoters) # 19242 - 31% of promoters
length(promoter_overlaps_MAX_peaks4)
length(promoter_overlaps_MAX_peaks4)/length(gencode_promoters)

```

This raises the question: Are these the same promoters?

# Overlap of promoters between replicates
```{r overlap of overlaps}

same_prom_ov12 <- findOverlaps(promoter_overlaps_MAX_peaks1,promoter_overlaps_MAX_peaks2)
length(same_prom_ov12) # 7525
mean_p12 = (length(promoter_overlaps_MAX_peaks1)+length(promoter_overlaps_MAX_peaks1))/2
mean_p12 # 13801 - average # of promoter peaks
length(same_prom_ov12)/mean_p12 # 54.5% overlap
length(unique(same_prom_ov12@to)) # 7462
length(unique(same_prom_ov12@to))/length(same_prom_ov12) # 99.2% are unique

same_prom_ov13 <- findOverlaps(promoter_overlaps_MAX_peaks1,promoter_overlaps_MAX_peaks3)
length(same_prom_ov13) # 13659
mean_p13 = (length(promoter_overlaps_MAX_peaks1)+length(promoter_overlaps_MAX_peaks3))/2
mean_p13 # 15186
length(same_prom_ov13)/mean_p13 # 89.9% overlap
length(unique(same_prom_ov13@to)) # 10044
length(unique(same_prom_ov13@to))/length(same_prom_ov13) # 73.5% are unique

same_prom_ov14 <- findOverlaps(promoter_overlaps_MAX_peaks1,promoter_overlaps_MAX_peaks4)
length(same_prom_ov14) # 13713
mean_p14 = (length(promoter_overlaps_MAX_peaks1)+length(promoter_overlaps_MAX_peaks4))/2
mean_p14 # 16521.5
length(same_prom_ov14)/mean_p14 # 83.0% overlap
length(unique(same_prom_ov14@to)) # 10055
length(unique(same_prom_ov14@to))/length(same_prom_ov14) # 73.3% are unique

same_prom_ov23 <- findOverlaps(promoter_overlaps_MAX_peaks2,promoter_overlaps_MAX_peaks3)
length(same_prom_ov23) # 7773
mean_p23 = (length(promoter_overlaps_MAX_peaks2)+length(promoter_overlaps_MAX_peaks3))/2
mean_p23 # 12178.5
length(same_prom_ov23)/mean_p23 # 63.8% overlap
length(unique(same_prom_ov23@to)) # 6483
length(unique(same_prom_ov23@to))/length(same_prom_ov23) # 83.4% are unique

same_prom_ov24 <- findOverlaps(promoter_overlaps_MAX_peaks2,promoter_overlaps_MAX_peaks4)
length(same_prom_ov24) # 7778
mean_p24 = (length(promoter_overlaps_MAX_peaks2)+length(promoter_overlaps_MAX_peaks4))/2
mean_p24 # 13514
length(same_prom_ov24)/mean_p24 # 57.5% overlap
length(unique(same_prom_ov24@to)) # 6484
length(unique(same_prom_ov24@to))/length(same_prom_ov24) # 83.4% are unique

same_prom_ov34 <- findOverlaps(promoter_overlaps_MAX_peaks3,promoter_overlaps_MAX_peaks4)
length(same_prom_ov34) # 16147
mean_p34 = (length(promoter_overlaps_MAX_peaks3)+length(promoter_overlaps_MAX_peaks4))/2
mean_p34 # 17906.5
length(same_prom_ov34)/mean_p34 # 90.1%
length(unique(same_prom_ov34@to)) # 15641
length(unique(same_prom_ov34@to))/length(same_prom_ov34) # 96.9% are unique

```
# Percentage of consensus peaks that overlap promoters?

```{r Consensus peaks}

cp12 <-subsetByOverlaps(peaks1,peaks2) 
length(cp12) # 11456
mean_12 = (length(peaks1)+length(peaks2))/2
mean_12 # 22285
length(cp12)/mean_12 # 51.4%
cp12_p = subsetByOverlaps(cp12, gencode_promoters) 
length(cp12_p) # 7020
length(cp12_p)/length(cp12) # 61.2%

cp13 <- subsetByOverlaps(peaks1,peaks3)
length(cp13) # 29123
mean_13 = (length(peaks1)+length(peaks3))/2
mean_13 # 57334
length(cp13)/mean_13 # 50.8%
cp13_p = subsetByOverlaps(cp13, gencode_promoters) 
length(cp13_p) # 13659
length(cp13_p)/length(cp13) # 46.9%

cp14 <- subsetByOverlaps(peaks1,peaks4)
length(cp14) # 29587
mean_14 = (length(peaks1)+length(peaks4))/2
mean_14 # 72532.5
length(cp14)/mean_14 # 40.8%
cp14_p = subsetByOverlaps(cp14, gencode_promoters) 
length(cp14_p) # 13709
length(cp14_p)/length(cp14) # 46.3%

cp23 <-subsetByOverlaps(peaks2,peaks3)
length(cp23) # 12853
mean_23 = (length(peaks2)+length(peaks3))/2
mean_23 # 48097
length(cp23)/mean_23 # 26.7%
cp23_p = subsetByOverlaps(cp23, gencode_promoters) 
length(cp23_p) # 7772
length(cp23_p)/length(cp23) # 60.5%

cp24 <-subsetByOverlaps(peaks2,peaks4) 
length(cp24) # 12870
mean_24 = (length(peaks2)+length(peaks4))/2
mean_24 # 63295.5
length(cp24)/mean_24 # 20.3%
cp24_p = subsetByOverlaps(cp24, gencode_promoters) 
length(cp24_p) # 7774
length(cp24_p)/length(cp24) # 60.4%

cp34 <-subsetByOverlaps(peaks3,peaks4)
length(cp34) # 72759
mean_34 = (length(peaks3)+length(peaks4))/2
mean_34 # 98344.5
length(cp34)/mean_34 # 74%
cp34_p = subsetByOverlaps(cp34, gencode_promoters) 
length(cp34_p) # 15996
length(cp34_p)/length(cp34) # 21.9%


```

cool so we saw before that it was 12,961 promoter overlaps 
now we have 12,331 which is pretty close 

# Do we dare to compare the promters ov from both methods?
# Go on then :)

```{r promoter overlaps between methods}

# so we previously made overlaps this way (changed to subsetByOverlaps)

subset_overlaps_peaks <- subsetByOverlaps(promoter_overlaps_pol2_peaks1, promoter_overlaps_pol2_peaks2)
subset_overlaps_peaks
# 12,257 unique overlap


# let's overlap this with peak overlaps

both_method_ov <- subsetByOverlaps(subset_overlaps_peaks, pol2_consensus_peaks_ov_promoters )
both_method_ov
# 12,257 that is a good sign that we got the same result !

# The one downside is we can't see how this all worked with @to and @from
both_method_ov@from 
#error :( 

```

Ok, that was a lot of overlapping. To summarize:
1) FindOverlaps will keep the @from and @to information 
2) Subset by overlaps only keeps the ranges that overlap 
3) you can do the same by indexing the original query or subject file
e.g. peaks1[unique(promoter_overlaps_pol2_peaks1@from)] 
4) Subset by overlaps is great for continuing to do lots of overlaps
5) @from = subject, @to = query


#########################
Excercise 
#########################

1) Find the peaks that overlap in all the replicates of your favorite protein.
- make sure to document how many peaks were in each starting file for overlaps
- how did this number change

2) How many of those peaks overlap promoters? 
- how many unqiue peaks are there that resulted in N overlaps?

3) Does the order of overlapping matter?

simple solution

```{r}

# create file path to load from
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/<your_folder>"
path <- "/CLASS_2023/CLASSES/03_Nextflow/00_my_chipseq/results/bwa/mergedLibrary/macs/broadPeak"

# load in each replicate
EP300_R1 <- rtracklayer::import(file.path(basepath,path, "EP300_R1_peaks.broadPeak"))
EP300_R2 <- rtracklayer::import(file.path(basepath,path, "EP300_R2_peaks.broadPeak"))
EP300_R3 <- rtracklayer::import(file.path(basepath,path, "EP300_R3_peaks.broadPeak"))
EP300_R4 <- rtracklayer::import(file.path(basepath,path, "EP300_R4_peaks.broadPeak"))

# finding length for each peak file
length(EP300_R1)
# 62260
length(EP300_R2)
# 66416
length(EP300_R3)
# 14319
length(EP300_R4)
# 16872


# iterative find overlaps
EP300_ov1 <- subsetByOverlaps(EP300_R1, EP300_R2) 
#38,409 overlaps
EP300_ov2 <- subsetByOverlaps(EP300_ov1, EP300_R3)
# 11,020 (got much smaller since R3 has only 14K peaks)
EP300_ov3 <- subsetByOverlaps(EP300_ov2, EP300_R4)
# 8,160 (even smaller since we reduced to 11K above)


# Does the order matter?

EP300_ov1 <- subsetByOverlaps(EP300_R3, EP300_R4) 
EP300_ov2 <- subsetByOverlaps(EP300_ov1, EP300_R1)
EP300_ov3 <- subsetByOverlaps(EP300_ov2, EP300_R2)
# now we have 8,199 !!

# This is an issue of reproducibility. If the order matters 
# then someone has to do the exact same order.
# not that much of a difference but not ideal.

```


Advanced Solution to prevent order mattering

```{r get a list of files}

#TODO? Does subset by overlaps take largest width? 

# make a list of all the peak file paths
peaks_fl <- list.files("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/CLASSES/03_Nextflow/00_my_chipseq/results/bwa/mergedLibrary/macs/broadPeak", full.names = T, pattern = ".broadPeak")


# extract the DBP name from the file path with regex
dbp <- sapply(peaks_fl, function(x){
  y <-  str_extract(x, "([^\\/]+$)")
  paste(unlist(strsplit(y, "_"))[c(1,2)], collapse = "_")
})

# let's see what peaks_fl looks like
peaks_fl

# adding names as dbp name and replicate
names(peaks_fl) <- dbp

# now let's see -- cool names are there
peaks_fl

# lapply will load all the files in one at a time using rtracklayer::import
# this can take a min
peaks_list <- lapply(peaks_fl, rtracklayer::import)

# checking we have names
names(peaks_list)

# let's find the number of peaks in each file
num_peaks <- sapply(peaks_list, length) %>% as.data.frame()

# creating a list of unique DBP names
dbps <- sapply(names(peaks_list), function(x){
  unlist(strsplit(x, "_"))[[1]]
}) %>%
  unique()

# let's check that dbps has unique names
dbps


# Consensus peaks from reduced "all peaks" set
# setting up function
consensus_from_reduced <- function(dbp, peaks_list) {
  
# index into list of Granges for each replicate for each dbp
  dbp_peaks <- peaks_list[grepl(as.character(dbp), names(peaks_list))]
  
  # this will unlist the Granges for each of the dbps "unlist"
  # This makes a list of all the peaks in all the file
  suppressWarnings(all_peaks <- GenomicRanges::reduce(unlist(as(dbp_peaks, "GRangesList"))))
  
  # Make a matrix of all peaks and the cols will be the replicate for each DBP
  peak_exists <- matrix(NA, nrow = length(all_peaks), ncol = length(dbp_peaks))
  for(i in 1:length(dbp_peaks)) {
    suppressWarnings(peak_exists[,i] <- as.numeric(countOverlaps(all_peaks, dbp_peaks[[i]]) > 0))
  }
  # filter to consensus requiring peaks to be in all replicates
  dbp_consensus <- all_peaks[rowSums(peak_exists) == ncol(peak_exists)]
  # Required only two replicates
  # dbp_consensus <- all_peaks[rowSums(peak_exists) > 1]
  return(dbp_consensus)
}

# Now run consensus_from_reduced on peak_list
# parameters: dbps (unique dbp names, function = consensu_from_reduced, peaks_list of Granges)
consensus_list <- lapply(dbps, consensus_from_reduced, peaks_list)

# adding names to results
names(consensus_list) <- dbps

# determining then length of resulting consensus peaks
sapply(consensus_list, length)
sapply(peak_list, length)

# now we can export each as .bed file
for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], 
paste0("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/CLASSES/04_R_basics/results/consensus_peaks/", names(consensus_list)[i], "_consensus_peaks.bed")) }


#TODO is there an lapply that can be done? 

